### ISSUE TITLE
Suggestion: Add "expecting version" field for app, to allow changing defaults.

### ISSUE THREAD

**Issue by niphlod** [web2py/web2py#611]
_From [i.like.p...@gmail.com](https://code.google.com/u/110204180798759823131/) on March 09, 2012 07:30:09_

web2py has maintained excellent backwards compatibility so far. However, sometimes, behavior SHOULD be changed. e.g., the response.generic_patterns default was changed because of a security problem, and this caused applications to break.

Suggested solution: add a call

web2py_configuration(version, ...) to the application's __init__.py file which states the assumptions of this version, e.g.

web2py_configuration(version=(1,99,4))

or alternatively a to a new CONFIG file int he web2py/applications/appname directory.

If there has been a breaking change, web2py could revert to old behavior and give a warning, or abort if there is no possible way to do that (or undesirable to do that).

e.g.

    if global.web2py_configuration.version \< (1,94.1):
        warn("Generic responses introduce a vulnerability. Please upgrade your app, notes at: http://web2py.com/upgrade-through-1-94-1" )
        if global.web2py_configuration.if_vulnerable == 'Do not abort, I know what I am doing and take full responsibility':  response.generic_patterns = ['*']
        else: raise Exception('Need changes, see warnings')

(an example that would let the user ignore vulnerabilities by an elaborate setting; might not be a good idea).

Now, make the wizard insert the current version number in the __init__.py or CONFIG file; Refuse running if there is no version information (or alternatively, warn about it and assume 1.99.7). Make the user edit the file manually (or through the app admin) when they have done the upgrade.

Breaking changes I would *like* to see:

- default sqlite setting to "pragma foreign_keys=1;" to enforce foreign key constraints. (I set it in my apps, but I think it should be default)
- default ondelete_action changed to 'RESTRICT' (or at least selectable). I personally set it to restrict always, because on a cross-linked database model, it is easy to accidentally cascade-delete millions of records when you are only trying to delete one.
- default for migration is off. migration can be enabled like today by changing the migrate=... flag, or through the admin app (in which case it turns off automatically after a transaction with a successful migration has occured)
- experimental features need to be turned on explicitly (e.g. a "require_experimental=['feature1', 'feature2', ...] in web2py_configuration)
- optionally switching db query result set from list() to generator/iterator.

You might not agree with about these changes, but putting infrastructure for future-proofing is a good idea regardless of the specific breaking changes.

_Original issue: http://code.google.com/p/web2py/issues/detail?id=702_

**Comment by niphlod** [comment]
_From [ABasta...@gmail.com](https://code.google.com/u/103720714265511973715/) on March 09, 2012 06:15:31_

This seems like it could get fairly complicated and difficult to maintain, both at the framework level and the app level.

\> - default ondelete_action changed to 'RESTRICT' (or at least selectable).

I think CASCADE is a more appropriate default. What do you mean by "selectable"?

\> - experimental features need to be turned on explicitly (e.g. a "require_experimental=['feature1', 'feature2', ...] in web2py_configuration)

Experimental features never break backward compatibility, they are just there if you decide to use them -- so why would you need to explicitly turn them on? Just don't use them if you don't need them.

\> - optionally switching db query result set from list() to generator/iterator.

If this would be optional, why would it have to break backward compatibility? Are you suggesting the default should be a generator?

**Comment by niphlod** [comment]
_From [i.like.p...@gmail.com](https://code.google.com/u/110204180798759823131/) on March 09, 2012 09:09:07_

\> This seems like it could get fairly complicated and difficult to maintain, both at the framework level and the app level.

At the app level, there is nothing to do accept occasionally modify the version file to the latest version, thereby saying: "I've updated my file after updating web2py, and I took care of all incompatiblities".

At the framework level, this is a just a programmatic list of (version, message) from the changelog. As long as we don't have multiple advancing branches, this is pretty simple.

\>  I think CASCADE is a more appropriate default. What do you mean by "selectable"?

I mean e.g. supporting something like:

   db = DAL("sqlite:blahbblah", adapter_args={...}, migrate=..., ondelete_action='SET NULL')

So you can set a default like you do with migrate, instead of having to repeat it for every table.

\> Experimental features never break backward compatibility, they are just there if you decide to use them -- so why would you need to explicitly turn them on? Just don't use them if you don't need them.

They break backward compatibility the other way around: if you used an experimental feature, and it was incompatibly changed or removed at a later version, when you update web2py, your app will break. By requiring activation, web2py will be able to give a meaningful post-upgrade warning ("the experimental feature your app depends on has been removed"). Also, it brings awareness when you start using it that it is experimental and might go away in a later version.

\> If this would be optional, why would it have to break backward compatibility? Are you suggesting the default should be a generator?

Yes, I am suggesting generator be a default, at least under some circumstances. I meant "optional" the other way around: You would still be able to switch it back if you wanted.

**Comment by niphlod** [comment]
_From [ABasta...@gmail.com](https://code.google.com/u/103720714265511973715/) on March 09, 2012 10:51:28_

\> At the app level, there is nothing to do accept occasionally modify the version file to the latest version, thereby saying: "I've updated my file after updating web2py, and I took care of all incompatiblities".

It looked like you were also proposing the ability to specify manual overrides via settings that would have to be coded and maintained.

Also, what happens if you want to accept some backward incompatible changes of a given version but not others?

\> At the framework level, this is a just a programmatic list of (version, message) from the changelog. As long as we don't have multiple advancing branches, this is pretty simple.

Your original post said, "If there has been a breaking change, web2py could revert to old behavior..." I assume that means when web2py detects on old version requirement, it has to execute the old behavior rather than the new. That involves more than just a programmatic list of (version, message) from the changelog (and gets more and more complicated as multiple backward incompatible changes are made to the same functionality over time -- all the old behaviors have to be maintained). And failing to do that would break backward compatibility, which we don't want to do.

\> I mean e.g. supporting something like:

\>   db = DAL("sqlite:blahbblah", adapter_args={...}, migrate=..., ondelete_action='SET NULL')

That seems reasonable, but I don't think it requires a breaking change (just make the default ondelete_action='CASCADE').

\> By requiring activation, web2py will be able to give a meaningful post-upgrade warning ("the experimental feature your app depends on has been removed").

I still don't think we should force users to explicitly turn on experimental features. At most, we could allow an optional registering of experimental features being used so a warning could be issued upon updates that affect the experimental feature.

Anyway, implementing an experimental feature turn-on requirement would not itself be a breaking change as long as we don't require it for any of the current experimental features (no promises have been made about how future experimental features would be implemented).
